tests/test_responses.py

```diff
@@ -0,0 +1,380 @@
+        "javascript": True,
+        "echoData": {"some_value": "here"},
+        "httpResponseHeaders": [
+            {"name": "Content-Type", "value": "text/html"},
+            {"name": "Content-Length", "value": len(PAGE_CONTENT)},
+        ],
+    }
+
+
+def api_response_body():
+    return {
+        "url": "https://example.com",
+        "httpResponseBody": b64encode(PAGE_CONTENT.encode("utf-8")),
+        "echoData": {"some_value": "here"},
+        "httpResponseHeaders": [
+            {"name": "Content-Type", "value": "text/html"},
+            {"name": "Content-Length", "value": len(PAGE_CONTENT)},
+        ],
+    }
+
+
+EXPECTED_HEADERS = {b"Content-Type": [b"text/html"], b"Content-Length": [b"44"]}
+EXPECTED_BODY = PAGE_CONTENT.encode("utf-8")
+
+
+@pytest.mark.parametrize(
+    "api_response,cls",
+    [
+        (api_response_browser, ZyteAPITextResponse),
+        (api_response_body, ZyteAPIResponse),
+    ],
+)
+def test_init(api_response, cls):
+    response = cls(URL, zyte_api=api_response())
+    assert response.zyte_api == api_response()
+
+    assert response.url == URL
+    assert response.status == 200
+    assert not response.headers
+    assert response.body == b""
+    assert not response.flags
+    assert response.request is None
+    assert response.certificate is None
+    assert response.ip_address is None
+    assert response.protocol is None
+
+
+@pytest.mark.parametrize(
+    "api_response,cls",
+    [
+        (api_response_browser, ZyteAPITextResponse),
+        (api_response_body, ZyteAPIResponse),
+    ],
+)
+def test_text_from_api_response(api_response, cls):
+    response = cls.from_api_response(api_response())
+    assert response.zyte_api == api_response()
+
+    assert response.url == URL
+    assert response.status == 200
+    assert response.headers == EXPECTED_HEADERS
+    assert response.body == EXPECTED_BODY
+    assert response.flags == ["zyte-api"]
+    assert response.request is None
+    assert response.certificate is None
+    assert response.ip_address is None
+    assert response.protocol is None
+
+
+@pytest.mark.parametrize(
+    "api_response,cls",
+    [
+        (api_response_browser, ZyteAPITextResponse),
+        (api_response_body, ZyteAPIResponse),
+    ],
+)
+def test_response_replace(api_response, cls):
+    orig_response = cls.from_api_response(api_response())
+
+    # It should still work the same way
+    new_response = orig_response.replace(status=404)
+    assert new_response.status == 404
+
+    new_response = orig_response.replace(url="https://new-example.com")
+    assert new_response.url == "https://new-example.com"
+
+    # Ensure that the Zyte API response is intact
+    assert new_response.zyte_api == api_response()
+
+    # The Zyte API response must also be fully replaceable
+    new_zyte_api = {"overridden": "value"}
+    new_response = orig_response.replace(zyte_api=new_zyte_api)
+    assert new_response.zyte_api == new_zyte_api
```

Comment:
What should be the behavior when the replaced zyte_api value contains one of the fields used to populate Response attributes? I.e. shouldn't e.g. response.replace(zyte_api={"httpResponseBody": "..."}) change response.body?

I think we should be explicit about it.

Reply:
It would probably be nice to make it smart that way. Maybe we can extract something out of `_process_response` to handle this well in an overriden `replace`. But we would also need to handle conflicts, e.g. `response.replace(zyte_api={"httpResponseBody": "..."}, body="…")`.

If we do not want to delay this pull request further, maybe we could make it so that `zyte_api` cannot be modified by `replace` (but we still make sure it is not lost either, and handle support to replace it in a follow-up pull request instead.

Reply:
This sounds fine to me - preserve .zyte_api on .replace calls, but raise an exception if user calls `response.replace(zyte_api=...)`?

Reply:
I'm not even sure what would be a use case for replacing .zyte_api attribute; it could be a feature we won't need for a long time.

Reply:
> What should be the behavior when the replaced zyte_api value contains one of the fields used to populate Response attributes? I.e. shouldn't e.g. response.replace(zyte_api={"httpResponseBody": "..."}) change response.body?

Great point! This could be tricky when Zyte API supports both `browserHtml` and `httpResponseBody` at the same time. How would `.replace()` work when both of them are present?

For now, we can be explicit in the test cases wherein performing `.replace()` using **zyte_api** should not result in the related response attribs to be updated. https://github.com/scrapy-plugins/scrapy-zyte-api/pull/10/commits/e0865e77e7a627b9a04b2638ca9e3241455d4421

Reply:
> but raise an exception if user calls response.replace(zyte_api=...)?

I realized, that I've pushed the commit without refreshing this page to see your new comments. 😅

Reply:
Not updating attributes is also not bad, but it seems an exception would be even better.

Reply:
Updated it to raise an exception in https://github.com/scrapy-plugins/scrapy-zyte-api/pull/10/commits/34a427f84fdd6f91d11579c8868d0ff2ccda0c5b